components/feature/pwa/src/test/java/mozilla/components/feature/pwa/ManifestStorageTest.kt

```diff
@@ -0,0 +1,94 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa
+
+import android.util.AtomicFile
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.engine.manifest.WebAppManifest
+import mozilla.components.concept.engine.manifest.WebAppManifestParser
+import mozilla.components.support.test.mock
+import mozilla.components.support.test.robolectric.testContext
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.doReturn
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.verify
+import java.io.FileOutputStream
+
+@ExperimentalCoroutinesApi
+@RunWith(AndroidJUnit4::class)
+class ManifestStorageTest {
+    @Test
+    fun `load returns null if entry does not exist or is invalid`() = runBlocking {
+        val storage = spy(ManifestStorage(testContext))
+        assertNull(storage.loadManifest("https://example.com"))
+
+        val mozFile = storage.mockFile("https://mozilla.org")
+        val fireFile = storage.mockFile("https://firefox.com")
+        `when`(mozFile.readFully()).thenReturn("[]".toByteArray())
+        `when`(fireFile.readFully()).thenReturn("{}".toByteArray())
+
+        assertNull(storage.loadManifest("https://mozilla.org"))
+        assertNull(storage.loadManifest("https://firefox.com"))
+    }
+
+    @Test
+    fun `load returns valid manifest`() = runBlocking {
+        val storage = spy(ManifestStorage(testContext))
+        val file = storage.mockFile("https://mozilla.org")
+        `when`(file.readFully())
+            .thenReturn("{\"name\":\"Mozilla\",\"start_url\":\"https://mozilla.org\"}".toByteArray())
+
+        val expected = WebAppManifest(name = "Mozilla", startUrl = "https://mozilla.org")
+        assertEquals(expected, storage.loadManifest("https://mozilla.org"))
+    }
+
+    @Test
+    fun `save saves the manifest as JSON`() = runBlocking {
+        val storage = spy(ManifestStorage(testContext))
+        val manifest = WebAppManifest(name = "Firefox", startUrl = "https://firefox.com")
+        val json = WebAppManifestParser().serialize(manifest)
+
+        val stream = mock<FileOutputStream>()
+        val file = storage.mockFile("https://firefox.com")
+        `when`(file.startWrite()).thenReturn(stream)
+
+        storage.saveManifest(manifest)
+        verify(stream).write(json.toString().toByteArray())
+        Unit
```

Comment:
Seems like good place to use `runBlockingTest {}`. :)

Reply:
Sadly `runBlockingTest` doesn't work with switching to `IO` for now.

Reply:
Do you mean `UncompletedCoroutinesError` in the end? I suppose we can pass `coroutineContext`/`dispatcher` into contructor to control execution context in tests.